File: shared/presentation/theme-ui/src/commonMain/kotlin/wallapp/pixel/theme/shape/AppShapes.kt
package wallapp.pixel.theme.shape

import androidx.compose.foundation.shape.CutCornerShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Shapes
import androidx.compose.ui.unit.dp
import wallapp.pixel.shape.ShapeStyle

object AppShapes {

    val CutCorner = Shapes(
        extraSmall = CutCornerShape(
            topStart = 8.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp
        ),
        small = CutCornerShape(
            topStart = 12.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp
        ),
        medium = CutCornerShape(
            topStart = 16.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp
        ),
        large = CutCornerShape(
            topStart = 20.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp
        ),
        extraLarge = CutCornerShape(
            topStart = 24.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp
        ),
    )

    val CutCorners = Shapes(
        extraSmall = CutCornerShape(
            topStart = 8.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 8.dp
        ),
        small = CutCornerShape(
            topStart = 12.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 12.dp
        ),
        medium = CutCornerShape(
            topStart = 16.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 16.dp
        ),
        large = CutCornerShape(
            topStart = 20.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 20.dp
        ),
        extraLarge = CutCornerShape(
            topStart = 24.dp,
            topEnd = 0.dp,
            bottomStart = 0.dp,
            bottomEnd = 24.dp
        ),
    )

    val RoundedCornerBottomStart = Shapes(
        extraSmall = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 4.dp,
            bottomEnd = 0.dp,
        ),
        small = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 8.dp,
            bottomEnd = 0.dp,
        ),
        medium = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 12.dp,
            bottomEnd = 0.dp,
        ),
        large = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 16.dp,
            bottomEnd = 0.dp,
        ),
        extraLarge = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 24.dp,
            bottomEnd = 0.dp,
        )
    )

    val RoundedCornersBottom = Shapes(
        extraSmall = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 4.dp,
            bottomEnd = 4.dp,
        ),
        small = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 8.dp,
            bottomEnd = 8.dp,
        ),
        medium = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 12.dp,
            bottomEnd = 12.dp,
        ),
        large = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 16.dp,
            bottomEnd = 16.dp,
        ),
        extraLarge = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 0.dp,
            bottomStart = 24.dp,
            bottomEnd = 24.dp,
        )
    )

    val RoundedCornersTop = Shapes(
        extraSmall = RoundedCornerShape(
            topStart = 4.dp,
            topEnd = 4.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp,
        ),
        small = RoundedCornerShape(
            topStart = 8.dp,
            topEnd = 8.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp,
        ),
        medium = RoundedCornerShape(
            topStart = 12.dp,
            topEnd = 12.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp,
        ),
        large = RoundedCornerShape(
            topStart = 16.dp,
            topEnd = 16.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp,
        ),
        extraLarge = RoundedCornerShape(
            topStart = 24.dp,
            topEnd = 24.dp,
            bottomStart = 0.dp,
            bottomEnd = 0.dp,
        )
    )


    val RoundedCorners = Shapes(
        extraSmall = RoundedCornerShape(4.dp),
        small = RoundedCornerShape(8.dp),
        medium = RoundedCornerShape(12.dp),
        large = RoundedCornerShape(16.dp),
        extraLarge = RoundedCornerShape(24.dp)
    )

    val RoundedCornersTopEndBottomStart = Shapes(
        extraSmall = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 4.dp,
            bottomStart = 4.dp,
            bottomEnd = 0.dp,
        ),
        small = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 8.dp,
            bottomStart = 8.dp,
            bottomEnd = 0.dp,
        ),
        medium = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 12.dp,
            bottomStart = 12.dp,
            bottomEnd = 0.dp,
        ),
        large = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 16.dp,
            bottomStart = 16.dp,
            bottomEnd = 0.dp,
        ),
        extraLarge = RoundedCornerShape(
            topStart = 0.dp,
            topEnd = 24.dp,
            bottomStart = 24.dp,
            bottomEnd = 0.dp,
        )
    )

    val Square = Shapes(
        extraSmall = CutCornerShape(0.dp),
        small = CutCornerShape(0.dp),
        medium = CutCornerShape(0.dp),
        large = CutCornerShape(0.dp),
        extraLarge = CutCornerShape(0.dp),
    )

    val Circle = Shapes(
        extraSmall = RoundedCornerShape(50),
        small = RoundedCornerShape(50),
        medium = RoundedCornerShape(50),
        large = RoundedCornerShape(50),
        extraLarge = RoundedCornerShape(50),
    )

    fun getShapes(shapeStyle: ShapeStyle): Shapes {
        return when (shapeStyle) {
            ShapeStyle.Circle -> Circle
            ShapeStyle.CutCorner -> CutCorner
            ShapeStyle.CutCorners -> CutCorners
            ShapeStyle.Rectangle -> Square
            ShapeStyle.RoundedCornerBottomStart -> RoundedCornerBottomStart
            ShapeStyle.RoundedCorners -> RoundedCorners
            ShapeStyle.RoundedCornersBottom -> RoundedCornersBottom
            ShapeStyle.RoundedCornersTop -> RoundedCornersTop
            ShapeStyle.RoundedCornersTopEndBottomStart -> RoundedCornersTopEndBottomStart
        }
    }
}

File: app/ios/wallApp/UI/Shapes/AppShapes.swift
//
//  AppShapes.swift
//  WallApp
//

import WallApp
import SwiftUI


struct AppShapes {
    
    static let cutCorner = ShapeDescriptors(
        extraSmall: ShapeDescriptor(topLeading: 8, shapeType: .cutCorner),
        small: ShapeDescriptor(topLeading: 12, shapeType: .cutCorner),
        medium: ShapeDescriptor(topLeading: 16, shapeType: .cutCorner),
        large: ShapeDescriptor(topLeading: 20, shapeType: .cutCorner),
        extraLarge: ShapeDescriptor(topLeading: 24, shapeType: .cutCorner)
    )
    
    static let cutCorners = ShapeDescriptors(
        extraSmall: ShapeDescriptor(topLeading: 8, bottomTrailing: 8, shapeType: .cutCorner),
        small: ShapeDescriptor(topLeading: 12, bottomTrailing: 12, shapeType: .cutCorner),
        medium: ShapeDescriptor(topLeading: 16, bottomTrailing: 16, shapeType: .cutCorner),
        large: ShapeDescriptor(topLeading: 20, bottomTrailing: 20, shapeType: .cutCorner),
        extraLarge: ShapeDescriptor(topLeading: 24, bottomTrailing: 24, shapeType: .cutCorner)
    )
    
    static let roundedCornerBottomStart = ShapeDescriptors(
        extraSmall: ShapeDescriptor(bottomLeading: 4, shapeType: .unevenRoundedRectangle),
        small: ShapeDescriptor(bottomLeading: 8, shapeType: .unevenRoundedRectangle),
        medium: ShapeDescriptor(bottomLeading: 12, shapeType: .unevenRoundedRectangle),
        large: ShapeDescriptor(bottomLeading: 16, shapeType: .unevenRoundedRectangle),
        extraLarge: ShapeDescriptor(bottomLeading: 24, shapeType: .unevenRoundedRectangle)
    )
    
    static let roundedCornersBottom = ShapeDescriptors(
        extraSmall: ShapeDescriptor(bottomLeading: 4, bottomTrailing: 4, shapeType: .unevenRoundedRectangle),
        small: ShapeDescriptor(bottomLeading: 8, bottomTrailing: 8, shapeType: .unevenRoundedRectangle),
        medium: ShapeDescriptor(bottomLeading: 12, bottomTrailing: 12, shapeType: .unevenRoundedRectangle),
        large: ShapeDescriptor(bottomLeading: 16, bottomTrailing: 16, shapeType: .unevenRoundedRectangle),
        extraLarge: ShapeDescriptor(bottomLeading: 24, bottomTrailing: 24, shapeType: .unevenRoundedRectangle)
    )
    
    static let roundedCornersTop = ShapeDescriptors(
        extraSmall: ShapeDescriptor(topLeading: 4, topTrailing: 4, shapeType: .unevenRoundedRectangle),
        small: ShapeDescriptor(topLeading: 8, topTrailing: 8, shapeType: .unevenRoundedRectangle),
        medium: ShapeDescriptor(topLeading: 12, topTrailing: 12, shapeType: .unevenRoundedRectangle),
        large: ShapeDescriptor(topLeading: 16, topTrailing: 16, shapeType: .unevenRoundedRectangle),
        extraLarge: ShapeDescriptor(topLeading: 24, topTrailing: 24, shapeType: .unevenRoundedRectangle)
    )
    
    static let roundedCorners = ShapeDescriptors(
        extraSmall: ShapeDescriptor(all: 4, shapeType: .roundedRectangle),
        small: ShapeDescriptor(all: 8, shapeType: .roundedRectangle),
        medium: ShapeDescriptor(all: 12, shapeType: .roundedRectangle),
        large: ShapeDescriptor(all: 16, shapeType: .roundedRectangle),
        extraLarge: ShapeDescriptor(all: 24, shapeType: .roundedRectangle)
    )
    
    static let roundedCornersTopEndBottomStart = ShapeDescriptors(
        extraSmall: ShapeDescriptor(topTrailing: 4, bottomLeading: 4, shapeType: .unevenRoundedRectangle),
        small: ShapeDescriptor(topTrailing: 8, bottomLeading: 8, shapeType: .unevenRoundedRectangle),
        medium: ShapeDescriptor(topTrailing: 12, bottomLeading: 12, shapeType: .unevenRoundedRectangle),
        large: ShapeDescriptor(topTrailing: 16, bottomLeading: 16, shapeType: .unevenRoundedRectangle),
        extraLarge: ShapeDescriptor(topTrailing: 24, bottomLeading: 24, shapeType: .unevenRoundedRectangle)
    )
    
    static let square = ShapeDescriptors(
        extraSmall: ShapeDescriptor(shapeType: .roundedRectangle),
        small: ShapeDescriptor(shapeType: .roundedRectangle),
        medium: ShapeDescriptor(shapeType: .roundedRectangle),
        large: ShapeDescriptor(shapeType: .roundedRectangle),
        extraLarge: ShapeDescriptor(shapeType: .roundedRectangle)
    )
    
    static let circle = ShapeDescriptors(
        extraSmall: ShapeDescriptor(shapeType: .circle),
        small: ShapeDescriptor(shapeType: .circle),
        medium: ShapeDescriptor(shapeType: .circle),
        large: ShapeDescriptor(shapeType: .circle),
        extraLarge: ShapeDescriptor(shapeType: .circle)
    )
    
    static func getShapes(shapeStyle: ClipShapeStyle) -> ShapeDescriptors {
        switch shapeStyle {
        case .cutCorner:
            return cutCorner
        case .cutCorners:
            return cutCorners
        case .roundedCornerBottomStart:
            return roundedCornerBottomStart
        case .roundedCorners:
            return roundedCorners
        case .roundedCornersBottom:
            return roundedCornersBottom
        case .roundedCornersTop:
            return roundedCornersTop
        case .roundedCornersTopEndBottomStart:
            return roundedCornersTopEndBottomStart
        case .rectangle:
            return square
        case .circle:
            return circle
        }
    }
}
